src/indexmap.jl

```diff
@@ -0,0 +1,239 @@
+LinearAlgebra.transpose(A::IndexMap) =
+    IndexMap(transpose(A.map), reverse(A.dims), A.cols, A.rows)
+LinearAlgebra.adjoint(A::IndexMap) =
+    IndexMap(adjoint(A.map), reverse(A.dims), A.cols, A.rows)
+
+
+#=
+combinations of IndexMap objects
+=#
+
+"""
+(possibly) simplified version of hcat - initial draft
+Requires 5-arg mul! to be efficient.
+"""
+function hcat_new(As::LinearMap...)
+    rows = collect(map(A -> size(A,1), As))
+    any(rows .!= rows[1]) && throw("row mismatch")
+    cols = collect(map(A -> size(A,2), As))
+    nmap = length(As)
+    col_offsets = [0; cumsum(cols)[1:nmap-1]]
+    dims = (rows[1], sum(cols))
+    T = promote_type(map(eltype, As)...)
+#   return +([LinearMap{T}(As[ii], dims ; offset=(0, col_offsets[ii]))
+#       for ii in 1:nmap]...)
+     return +([IndexMap{T}(As[ii], dims ; offset=(0, col_offsets[ii]))
+         for ii in 1:nmap]...)
+end
+
+
+"""
+`B = block_diag(As::LinearMap...)`
+
+Block diagonal `LinearMap`, akin to `SparseArrays.blockdiag()`
+Requires 5-arg mul! to be efficient.
+"""
+function block_diag(As::LinearMap...)
+    rows = collect(map(A -> size(A,1), As))
+    cols = collect(map(A -> size(A,2), As))
+    dims = (sum(rows), sum(cols))
+    nmap = length(As)
+    col_offsets = [0; cumsum(cols)[1:nmap-1]]
+    row_offsets = [0; cumsum(rows)[1:nmap-1]]
+    T = promote_type(map(eltype, As)...)
+    return +([IndexMap{T}(As[ii], dims ; offset=(row_offsets[ii], col_offsets[ii]))
+        for ii in 1:nmap]...)
+end
+
+
+#=
+multiplication with vectors
+
+When IndexMaps are put in a LinearCombination, hopefully A_mul_B!
+is called only on the "first" block, so that the overhead of `fill`
+happens just once.
+=#
+
+#if VERSION < v"1.3.0-alpha.115"
+
+"""
+Multiply a single IndexMap with a vector.
+Basic *(A,x) in LinearMaps.jl#L22 uses similar() so must zero `y` here.
+In principle we could zero out just the complement of `A.rows`.
+"""
+function A_mul_B!(y::AbstractVector, A::IndexMap, x::AbstractVector)
+    m, n = size(A)
+#    @show "todoIM-AB:", size(A)
+    @boundscheck (m == length(y) && n == length(x)) || throw(DimensionMismatch("A_mul_B!"))
+    fill!(y, zero(eltype(y)))
+    @views @inbounds mul!(y[A.rows], A.map, x[A.cols], true, true)
+    return y
+end
+
+
+"""
+This 5-arg mul! is based on LinearAlgebra.mul! in LinearMaps.jl
+with modifications to support the indexing used by an IndexMap.
+y = α B x + β y
+"""
+function LinearAlgebra.mul!(y_in::AbstractVector, B::IndexMap,
+        x_in::AbstractVector, α::Number=true, β::Number=false)
+    length(y_in) == size(B, 1) || throw(DimensionMismatch("mul!"))
+    A = B.map
+#    @show "todoIM:", size(B), α, β
+    xv = @views @inbounds x_in[B.cols]
+    yv = @views @inbounds y_in[B.rows]
+    if isone(α) # α = 1, so B x + β y
+        if iszero(β) # β = 0: basic y = B*x case
+            fill!(y_in, zero(eltype(y_in)))
+            A_mul_B!(yv, A, xv)
+        elseif isone(β) # β = 1
+            yv .+= A * xv
```

Comment:
This should be `mul!(yv, A, xv, true, true)`, otherwise the right hand side will allocate, even if `A` is matrix.